### ISSUE TITLE
Feature: Dynamic catalogs

### ISSUE THREAD

**Issue by mkavidas** [Stremio/stremio-addon-sdk#219]

**Comment by TRtomasz** [comment]
transfered this issue to stremio-addon-sdk repository
Any change that is applied to addon-sdk must also be then implemented in all Stremio apps so i don't think this will happen at all. Unless we decide to make another version of our addon sdk.

**Comment by jaruba** [comment]
@mkavidas The problem you are describing is not a problem with the addon specification, but a problem with auto-updating addon manifests. Currently, Stremio will only auto-update addon manifests for users if the addons are published to the official addon catalog and if those addons do not use any addon configurations. (such as configurations from the addon settings page, because making any such configuration changes the addon manifest url from the default one)

There is currently no ETA on changing how addon manifests are auto-updated, but it is on our radar for some time and considered a priority task.

**Comment by dblobster01** [comment]
@jaruba I disagree that the issue is manifest not auto updating. My understanding is the purpose of the manifest should be to define the functionality of an addon and not to be a source for what should be dynamic content. Anyways, having the catalogs being queried dynamically would allow for a lot of interesting functionality:

Dynamic catalog names (even per user).
Seasonal catalogs.
Randomizing the order that catalogs appear to make it easier to discover content/ keep things fresh. 
Catalogs themselves could be dynamically generated server side (again per user)

The only solution as of now would be to continuously recreate the manifest and then have users reinstall an addon (possible automatically if it’s official). 

I’m failing to see the importance of having catalogs defined in the manifest. 

Think about how most streaming services handle their “catalogs”. They generate a unique listing and order based on user, season, trends etc.

**Comment by mkavidas** [comment]
Would this be a feature that the Stremio community is at all interested in? I would be happy to work on it myself if so. 

My proposed solution would be:

1. leave existing catalog resource untouched to support backwards compatibility and simpler catalog design via manifest.json
2. Introduce a new optional resource type: catalogs. This resource will return the exact same information as required by catalog resource in the manifest. {name,id extras.}
3. In Stremio core we can have one step before we populate the catalog data where we combine the results of manifest.catalog and addon/catalogs, if duplicate ids exist prefer addon/catalogs. 
4. Shouldn't require much change for stremio clients as the endpoint for actual catalog data is the same /catalog/movie/id. All current addons will continue to function without requiring any updates.

On the SDK side we only need to add the endpoint addon/catalogs which returns the same information as would be in the manifest defined catalog. 

I realize that this is a naive approach that doesn't consider things like caching and other client side considerations that I am surely missing. Alternatively, if it is vital that the routes be pre-defined in the manifest, you could have all of the catalogs defined in the manifest (id is required) but the name of the catalog can be overridden by catalog/movie/id. This would allow an addon to alter the order in which catalogs appear and their content server side. 
 
Please let me know if this is feasible and if someone could provide some direction on where I should start working on this, let me know.

**Comment by bakes82** [comment]
> [@mkavidas](https://github.com/mkavidas) The problem you are describing is not a problem with the addon specification, but a problem with auto-updating addon manifests. Currently, Stremio will only auto-update addon manifests for users if the addons are published to the official addon catalog and if those addons do not use any addon configurations. (such as configurations from the addon settings page, because making any such configuration changes the addon manifest url from the default one)
> 
> There is currently no ETA on changing how addon manifests are auto-updated, but it is on our radar for some time and considered a priority task.

Has this ever been addressed, much like the OP. I have a custom addon with my own collections I generate and then have the manifest render dynamically, if I add/remove a collection the manifest will render with a new version, Streamio doesnt pick up the changes and the only way I can seem to get it to show the new/removed collections is to uninstall/reinstall the addon eventhough the manifest url is showing the correct new version with the new collection the streamio side still show the old version number and old collections, not even a notification that there needs to be an update.  Now if use the web ui and repaste the URL it will show an update button, IDT the windows client does that
